pandas/core/case_when.py

```diff
@@ -0,0 +1,151 @@
+        ``pd.NA`` for extension dtypes).
+
+    level : int, default None
+        Alignment level if needed.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when(df.a.gt(0), df.a,   # condition, replacement
+    ...              df.b.gt(0), df.b,
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
+    len_args = len(args)
+    if not len_args:
+        raise ValueError(
+            "Kindly provide at least one boolean condition, "
+            "with a corresponding replacement."
+        )
+    if len_args % 2:
+        raise ValueError(
+            "The number of boolean conditions should be equal "
+            "to the number of replacements. "
+            "However, the total number of conditions and replacements "
+            f"is {len(args)}, which is an odd number."
+        )
+
+    conditions = args[-2::-2]
+    replacements = args[-1::-2]
+    common_dtype = [infer_dtype_from(replacement)[0] for replacement in replacements]
+    if default is not None:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtype.append(arg_dtype)
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            construct_1d_arraylike_from_scalar(
+                value=replacement, length=len(condition), dtype=common_dtype
+            )
+            if is_scalar(replacement)
+            else replacement.astype(common_dtype, copy=False)
+            if isinstance(replacement, ABCSeries)
+            else pd_array(replacement, dtype=common_dtype, copy=False)
+            for condition, replacement in zip(conditions, replacements)
+        ]
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtype, copy=False)
+    else:
+        common_dtype = common_dtype[0]
+    # TODO: possibly extend this to a DataFrame?
+    if not isinstance(default, ABCSeries):
+        default_index = len(args[0])
+        for condition in conditions[::-1]:
+            if isinstance(condition, ABCSeries):
+                default_index = condition.index
+                break
```

Comment:
I didn't consider that the first condition could be a non-pandas object. With this, it seems to me this has a bit of an odd behavior of using the first indexed condition rather than always using the first one. In hindsight, maybe even my suggestion of using the first condition for the index is guessing at what the user wants to do, and is something we should avoid.

When `default` has no index, we could check that for those that are Series, the conditions and/or values have the same index. If they aren't all equal, we raise. In this case they are all equal, it's unambiguous what the resulting index should be. I worry this is an expensive check. But in what I think is the common case where a single DataFrame is used (e.g. `pd.case_when(df['a'] > 5, 10, df['b'] < 3, 5, default=0)`), this check is very cheap.

If we do go this route, we can also entertain adding an optional `index` argument to the function specifically to tell use what the user wants the resulting index to be. But this isn't necessary.

What do you think @samukweku? Also wouldn't mind getting a few more eyes on this, cc @mroeschke @jorisvandenbossche

Reply:
For a non-pandas, list-like condition, I would assume the "index" would be the same as the original object unless I'm not totally understanding the scenario

Reply:
@mroeschke - I'm guessing you're thinking `DataFrame.case_when`. This function is top-level: `pd.case_when`.

Reply:
Should we get rid of the top level function and just assign it to a Series/DataFrame instead?

Reply:
@samukweku - In the linked issue I've advocated that `Series.case_when` would be quite useful - but then you'd remove the `default` argument (the Series values are the default). However, that would not cover a use case like

    result = pd.case_when(df['a'] > 5, 1, df['b'] < 3, 2, default=0)

The Series alternative would be

    result = pd.Series(0, index=df.index).case_when(df['a'] > 5, 1, df['b'] < 3)

While I like the explicitness of the 2nd version and do prefer it, I can understand if others find the first more natural.

I don't know what `DataFrame.case_when` would do. Modify all columns? This doesn't seem like a common use case.